### PR TITLE
Implement tray icon

### LINKED ISSUE #1548: Close desktop version to system tray
As I understand - delta chat can not be "closed" to system tray? Or I missing this setting?

### LINKED ISSUE #1692: There is no app icon for i3wm/i3-gaps (i3blocks)
- Operating System (Linux/Mac/Windows/iOS/Android): Arch Linux
- Delta Chat Version: Latest. Installed from aur git.
- Expected behavior: Showing icon on the pannel i3blocks, as doing other apps, for example Telegram on the screenshot below.
- Actual behavior: No any icon
- Screenshots:
18h_15m_42s_30_May_2020

- Logs:

### PR DESCRIPTION
**Pull request by Jikstra**
- [x] Refactor code and remove if/else indentations if possible
- [x] Hide window when clicking on tray icon if it's visible
- [x] refactor code to not depend on ipcBackend
- [x] minimize window instead of hiding it, to make it faster to show it again
- [x] focus window if it's on another virtual desktop instead of hiding it
- [x] package-lock.json should not change?
- [x] in the menu it should be named `minimize to tray` instead of `reduce`
- [x] check behavior on mac (not the trayicon, just that it isn't closed fully and can be restored from the docs)
   - I just checked mac behavior and it does not look good. Clicking on the dock icon does not restore the window.
- [x] investigate gnome bug? (toggle tray):

- [x] move setting to experimental

about the gnome bug we could also create an issue when merging and postpone the investigation.

successor of 1840
closes #1692
closes #1548

### CODE REVIEW

**Comment by Jikstra** [comment]
Actually I would like to merge this into @pepea28/tray but somehow I can't make this work. @Simon-Laux @pepea28 any ideas?

**Comment by Simon-Laux** [comment]
you'll need access to @pepea28's repo, @Jikstra

**Comment by Simon-Laux** [comment]
otherwise we could make a collective tray icon branch on the deltachat/deltachat-desktop repo in which we merge the prs regarding it.

**Comment by Jikstra** [comment]
@Simon-Laux hmm but also for creating a pull request? Other thing in my mind, probably it's fine to not draw the badge icon for now and do it in a seperate pr. We need to do some canvas magic, all the libraries i saw are either not on npm or do weird things. element-web is also doing it by themselves.

**Comment by pepea28** [comment]
The settings part is not present in this PR. So it is difficult to test. Is it normal ?
It's a good idea to focus the window if it's on another virtual desktop.

Does minimizing the window instead of hiding it leave Deltachat in the taskbar ?

**Comment by Jikstra** [comment]
@pepea28 
> The settings part is not present in this PR. So it is difficult to test. Is it normal ?

Should work? Should work, at least I didn't remove it :D

> Does minimizing the window instead of hiding it leave Deltachat in the taskbar ?

Yes, this is not optimal but hiding and showing it feels to slow on my device, that's why I chose this path.

**Comment by Simon-Laux** [comment]
- [ ] in the menu it should be named `minimize to tray` instead of `reduce`
- [ ] check behavior on mac (not the trayicon, just that it isn't closed fully and can be restored from the docs)
- [ ] investigate gnome bug? (toggle tray):

File: src/main/menu.ts [hunk 1]
```diff
@@ -169,6 +169,13 @@ function getMenuTemplate(logHandler: LogHandler): rawMenuItem[] {
                 translate: 'global_menu_file_quit_desktop',
                 role: 'quit',
               },
+              {
+                translate: 'global_menu_file_reduce_desktop',
+                accelerator: 'Escape',
```
Comment (Simon-Laux) [inline src/main/menu.ts:174]: I wouldn't introduce the escape shortcut here, escape also does other things and I think the normal user gets confused when escape results in the window being invisible.

File: src/main/menu.ts [hunk 2]
```diff
@@ -169,6 +169,13 @@ function getMenuTemplate(logHandler: LogHandler): rawMenuItem[] {
                 translate: 'global_menu_file_quit_desktop',
                 role: 'quit',
               },
+              {
```
Comment (Simon-Laux) [inline src/main/menu.ts:172]: This option should only be shown when the tray icon setting is active

File: src/main/tray.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      type: 'normal',
+      click() {
+        win.show()
+      },
+    },
+    {
+      id: 'reduce_window',
+      label: tx('global_menu_file_reduce_desktop'),
+      type: 'normal',
+      accelerator: 'Escape',
```
Comment (Simon-Laux) [inline src/main/tray.ts:58]: I don't know what the `accelerator` should do here, I suggest to remove this line

File: src/main/menu.ts (hunk 2 above)
Comment (Simon-Laux) [reply src/main/menu.ts:172]: or if we are on mac

**Comment by Jikstra** [comment]
Just for completeness, the electron bug for the multiple tray icons:
https://github.com/electron/electron/issues/17622

**Comment by Jikstra** [comment]
@Simon-Laux what do you mean with "I just checked mac behavior and it does not look good. Clicking on the dock icon does not restore the window." ? Just tried it out on the one around me, activating the tray icon has no effect (and could get hidden) but besides this everything works as expected.

**Pull request merged**